libclamav_rust/src/codesign.rs

```diff
@@ -0,0 +1,552 @@
+        let root_ca_bytes = std::fs::read(root_ca_path)?;
+        let root_ca: X509 = X509::from_pem(&root_ca_bytes)?;
+
+        debug!("Root CA: {:?}", root_ca);
+
+        Ok(Verifier { root_ca })
+    }
+
+    pub fn verify(&self, data: &[u8], pkcs7: &Pkcs7) -> Result<String, Error> {
+        if let Some(signed) = pkcs7.signed() {
+            if let Some(cert_stack) = signed.certificates() {
+                let root_ca_serial = self.root_ca.serial_number().to_bn()?;
+                debug!("Checking if the root CA's serial matches any in the signature's certificate stack...");
+                let mut top_level_signer: Option<String> = None;
+
+                // Check each cert in the pkcs7 cert stack to see if it matches the root CA.
+                // If we can't find a matching serial number, then we can't verify the pkcs7 signature.
+                // That doesn't mean the signature is invalid, only that we don't have the required public key to verify it.
+                for cert in cert_stack {
+                    let serial = cert.serial_number().to_bn()?;
+                    if top_level_signer == None {
+                        let signer = cert
+                            .subject_name()
+                            .entries()
+                            .find(|name_entry| {
+                                name_entry.object().nid() == openssl::nid::Nid::COMMONNAME
+                            })
+                            .ok_or(Error::InvalidDigitalSignature(
+                                "Certificate in the signature's cert stack does not have a Common Name entry".to_string(),
+                            ))?
+                            .data()
+                            .as_utf8()?
+                            .to_string();
+                        debug!("Top level signer serial: {}", signer);
+                        top_level_signer = Some(signer);
+                    }
+
+                    if root_ca_serial == serial {
+                        // found a matching serial number in the pkcs7 cert stack matching the provided root CA.
+                        // We can verify the signature.
+                        debug!("Certificate serial is a match: {:?}. We should be able to use this root CA to verify this signature.", serial.to_dec_str()?);
+
+                        // create store with root CA
+                        let mut store_builder = X509StoreBuilder::new()?;
+                        store_builder.add_cert(self.root_ca.clone())?;
+                        let store = store_builder.build();
+
+                        // verify signature
+                        let certs = stack::Stack::new()?;
+                        let flags = Pkcs7Flags::DETACHED | Pkcs7Flags::BINARY;
+                        let mut output = Vec::new();
+
+                        let result =
+                            pkcs7.verify(&certs, &store, Some(data), Some(&mut output), flags);
```

Comment:
Need to make sure CRL / OSCP verification is not performed in the initial release, and to set up our actual signing kes/certs have that stuff set up for future use.